### ISSUE TITLE
[RFC] Selective logits projection for paged text prefill

### ISSUE THREAD

**Issue by HermanZeng** [vllm-project/vllm-metal#525]
## Motivation

The paged text path packs all scheduled tokens into one sequence in this order:

```text
[all decode/spec-verification rows][all rows of prefill 0][all rows of prefill 1]...
```

The target forward currently materializes logits for every packed row. Prefill
sampling then keeps only the final row of each prefill segment, while decode and
speculative verification consume the decode prefix:

- [`model_runner.py`: the target forward produces full logits](https://github.com/vllm-project/vllm-metal/blob/5f4e50d1fcff299b9668177e7fba97955853437e/vllm_metal/v1/model_runner.py#L1097-L1104)
- [`sampling_batch.py`: prefill sampling extracts only each segment's last row](https://github.com/vllm-project/vllm-metal/blob/5f4e50d1fcff299b9668177e7fba97955853437e/vllm_metal/v1/sampling_batch.py#L436-L499)
- [`structured_output.py`: grammar masking targets the same decode spans and prefill-final rows](https://github.com/vllm-project/vllm-metal/blob/5f4e50d1fcff299b9668177e7fba97955853437e/vllm_metal/v1/structured_output.py#L136-L187)

For a prefill-heavy batch this constructs a tensor of shape
`[1, total_scheduled_tokens, vocab_size]` although only
`num_decode_tokens + num_prefill_requests` rows are currently consumed.

This was already identified as a separate follow-up in
[#482](https://github.com/vllm-project/vllm-metal/issues/482):

> prefill logits are materialized for every position before row extraction
> repo-wide — both the target forward and the draft ingest do this; that's its
> own optimization issue

I could not find a standalone issue or pull request implementing that follow-up
as of 2026-07-20.

I am filing this RFC before writing an implementation to ask whether maintainers
consider the optimization worthwhile and whether the proposed boundary fits the
project. The design below is a starting point for discussion, not an assumed
decision.

Upstream vLLM uses a relevant high-level contract: it computes a
one-dimensional `logits_indices`, selects the required hidden-state rows, and
only then calls the model-owned logits projection:

```python
sample_hidden_states = hidden_states[logits_indices]
logits = self.model.compute_logits(sample_hidden_states)
```

See [`gpu_model_runner.py`](https://github.com/vllm-project/vllm/blob/e6d1310b2ac52e56e266450693af9de12a871a51/vllm/v1/worker/gpu_model_runner.py#L4428-L4429).

### Local evidence (isolated LM-head microbenchmark)

On an M1 Pro with `Qwen/Qwen3-0.6B` (`vocab_size=151,936`, BF16), projecting
random hidden states through the real tied output head produced the following.
The “current pattern” evaluates only a last-row slice of the full-head result,
matching the lazy graph shape used by the serving path; the timings therefore
also confirm that MLX does not push this slice through the output projection.

| Input rows | Current: full head, then evaluate last-row slice | Proposed: select hidden row, then project | Logical full logits size |
|---:|---:|---:|---:|
| 512 | 67.19 ms | 2.12 ms | 148.4 MiB |
| 1,024 | 140.68 ms | 2.51 ms | 296.8 MiB |
| 2,048 | 268.62 ms | 2.64 ms | 593.5 MiB |

These numbers isolate the output projection; they are **not** an end-to-end
serving speedup claim. A performance PR still needs the before/after serving
benchmarks required by `docs/CONTRIBUTING.md`.

Selecting rows before projection can choose a different MLX matmul kernel or
reduction shape. On the same random-hidden-state tests, the selected-row logits
differed from the corresponding full-projection row by at most `0.0078125` to
`0.015625`; a short real-prompt forward observed `0.03125`. The argmax was
unchanged in all of these checks. Correctness validation should therefore use
dtype-aware logit tolerances plus deterministic token parity, not require
bitwise equality of every logit.

### 1. Proposed row-selection contract

For the existing packed order, required original rows are:

```python
logits_indices = [
    *range(num_decode_tokens),
    *(token_cu_seqlens[num_decode_segments + i + 1] - 1
      for i in range(len(prefill_reqs))),
]
```

This preserves every row needed by plain decode and speculative verification,
then appends exactly one last-token row per prefill segment. The resulting
compact logits order is deterministic:

```text
[unchanged decode/spec rows][prefill 0 final row][prefill 1 final row]...
```

No general row-remapping object is needed.

I propose bypassing the optimization when `len(logits_indices)` equals the number
of input rows, because pure decode already needs every row and would gain
nothing.

### 2. Proposed `ModelAdapter.target_forward()` extension

One possible minimal implementation is to add an optional `logits_indices`
argument. When selective projection is known to preserve a model family's
wrapper semantics, the adapter would:

1. run the model backbone to obtain all hidden states (attention/KV-cache work
   remains unchanged);
2. select the requested hidden-state rows with `mx.take`;
3. apply the model-owned output projection and all model-specific logit
   post-processing only to those rows;
4. return the full row-major hidden states only when MTP requests them.

The forward result could expose one small piece of metadata, for example
`logits_indices_applied: bool`, so the runner can distinguish compact output
from a compatibility fallback without guessing from shapes.

When the adapter cannot prove equivalence with the model wrapper, it would call
the existing `model(input_ids, cache=...)` path and return full logits.

This gate matters. `mlx-lm` does not currently expose a uniform
`compute_logits(hidden_states)` contract across every model family. Examples
include Cohere's `logit_scale`, Granite's `logits_scaling`, Falcon H1's head
multiplier, and several soft-cap variants. The existing private
`_compute_target_logits()` helper was added for Gemma4 MTP, so I do not think it
should be treated as universal without model-family parity tests.

I suggest enabling the first implementation only for a small, explicitly tested
set of text model families (starting with Qwen3) and retaining transparent full
logits fallback for the rest. A future `mlx-lm` model-owned `compute_logits`
contract could remove that allow-list.

### 3. Proposed paged-forward layout metadata

- `token_cu_seqlens`: the existing boundaries in the original packed hidden
  states. Gemma4 MTP continues to use these when selecting target hidden rows.
- `logits_cu_seqlens`: boundaries in the returned logits tensor. For compact
  output, decode/spec spans retain their existing lengths and each prefill span
  has length one. For a fallback full output, this equals
  `token_cu_seqlens`.

The runner would pass `logits_cu_seqlens` to prefill sampling and
structured-output masking, and `token_cu_seqlens` to the MTP proposer. Their
existing row-selection logic would then remain valid without a new layout
abstraction.

Because the entire decode/spec prefix is preserved in place:

- `PagedDecodeSegment.start_row` remains valid;
- greedy speculative verification needs no remapping;
- plain decode sampling needs no remapping;
- grammar row spans for decode/spec remain valid.

### 4. Possible initial scope

Proposed in scope:

- standard paged text target forward;
- plain prefill and mixed prefill/decode batches;
- existing speculative-verification rows;
- existing structured-output row targeting;
- Gemma4 MTP with full target hidden states retained;
- one selected row for every scheduled prefill segment, including an
  intermediate chunk, to preserve current behavior exactly.

Proposed out of scope:

- multimodal forward;
- pipeline-parallel forward;
- pooling models;
- draft-model `_ingest_and_draft_first()` (same opportunity, separate change);
- prompt logprobs (currently unsupported on this path);
- skipping sampling for intermediate chunked-prefill requests;
- a generic `LogitsLayout`/registry or a new environment flag.

### Unit and integration correctness

- Index construction for pure prefill, pure decode, mixed batches, multiple
  prefill requests, and multi-row speculative verification.
- Compact row ordering and `logits_cu_seqlens` invariants.
- Structured-output masking for prefill, decode, and spec-decode row spans.
- Gemma4 MTP seeds continue to index the full target hidden-state layout.
- Model-adapter parity tests that cover tied embeddings, `lm_head`, and each
  supported logit post-processing rule.
- Dtype-aware selected-logit comparison against the current full projection,
  plus deterministic greedy token parity on representative models.
- Existing non-slow test suite and lint commands from `docs/CONTRIBUTING.md`.

### Performance

- Isolated LM-head benchmark reporting selected rows, vocabulary size, dtype,
  latency, and peak logits size.
- End-to-end `vllm bench serve` before/after with the documented sonnet setup,
  reporting TTFT, TPOT, and E2EL p50/p99.
- At least one long-prefill workload and one mixed prefill/decode workload.
- A pure-decode no-regression check; the selective path should be bypassed when
  every row is needed.

## Alternatives considered

1. **A generic `LogitsLayout` with original/compact row maps.** I do not propose
   this for the initial change: the existing packed order makes the mapping
   trivial, and the class would add state and synchronization points without
   enabling a current requirement.
2. **Slice after full logits projection.** This preserves the current API but
   does not remove the expensive `[tokens, vocab]` projection or allocation.
3. **Optimize target and draft ingest together.** I would keep draft ingest
   separate so review and performance attribution stay bounded. It could be
   considered independently if maintainers find the target-side direction
   worthwhile.
4. **Apply the split backbone/head path to every `mlx-lm` model immediately.**
   I do not propose this because model wrappers do not yet share a complete
   logits post-processing contract.

## Request for maintainer feedback

This issue is a design proposal, not an assumption that the change will be
accepted or that I own the implementation. Before writing a PR, I would like
maintainer feedback on:

1. whether this optimization is worth pursuing in vllm-metal;
2. whether the adapter-owned, model-family opt-in fallback is the right
   correctness boundary; and
3. whether Qwen3 is an appropriate initial scope.

If maintainers support the direction, I am interested in working on a narrowly
scoped first PR. I am also happy for maintainers to suggest a different design,
scope, or owner.

## Any Other Things

- I checked the open issue and pull-request lists immediately before filing and
  found no standalone duplicate. #482 is the related prior observation.
- **AI assistance disclosure:** This RFC draft was prepared with OpenAI Codex.
  The proposer independently reviewed the referenced code paths, ran the local
  measurements, and takes responsibility for the claims in this issue. Any
  future code contribution would be separately tested, reviewed, and disclosed.

**Comment by robertomeroni** [comment]
@HermanZeng This RFC is thorough and I should say up front that I built an implementation of it before
noticing it had gone three weeks without a reply, not to jump the queue, and I'm happy for you to own
it. Posting what I have in case it's useful either way, and to answer your three questions with numbers.

- Is it worth pursuing: yes on this hardware. Interleaved `vllm bench serve` A/B on an M3 Pro
(Qwen3-0.6B, 2048-token prompts, 40 requests, 5 iterations per arm alternating): **+15.8% output
throughput, −12.9% mean TTFT, −15.1% mean TPOT**, with the two arms' ranges fully disjoint on every
metric. So it holds up e2e, not just in the isolated head benchmark.

- Is the adapter-owned boundary right: I think so, and your contract worked unchanged; the runner
computes `logits_indices` because only it knows the paged layout, the adapter does the `mx.take` and the
projection because it owns the head. `logits_cu_seqlens` alongside `cu_seqlens` also worked exactly as
you described: logits consumers read the compact boundaries, the MTP proposer keeps reading the packed
ones.

- Is Qwen3 the right initial scope: this is where I diverged. Rather than an allowlist, the adapter
probes the loaded model once at startup (a one-token cacheless forward comparing the model's own output
against the split backbone/head path), and anything that mismatches keeps full logits. It costs one
forward at load and it covers Cohere/Granite/Falcon-H1 without anyone having to maintain a list against
`mlx-lm`. Note this probe compares a single row through both paths, so it tests *wrapper semantics*
rather than the selected-row kernel-numerics question you raised; that one I handled by comparing
selected rows against the corresponding full-projection rows in tests.

I also dropped the `logits_indices_applied` flag you proposed: with one `target_forward` implementation
it could never be false, so the fallback branch reading it was dead code.

Two things I'd value your or a maintainer's read on: whether the load-time probe is acceptable or an
explicit opt-in is preferred, and whether dropping non-last prefill rows is fine given prompt logprobs
are unimplemented on this path today (`prompt_logprobs_dict={}` is hardcoded).


AI assistance disclosure: Opus 5 can explain what I did better than me, so I used it to write this.

**Comment by now-ing** [comment]
Audited the current state of main against this RFC — the core proposal has been implemented by #590 (`db2906c`, "Project logits only for sampled rows on the paged prefill path") and looks complete for the RFC's stated scope:

- `_paged_logits_layout` (`vllm_metal/v1/model_runner.py`) selects every decode/spec row plus each prefill segment's last row, collapses the packed boundaries accordingly, and passes `logits_indices` into the adapter, where `mx.take` gathers exactly those rows before the LM head.
- The no-op path is preserved: pure-decode steps (or all-single-token prefill segments) skip the gather entirely and keep the model's own `__call__`.
- Adoption is gated (`_selective_logits_supported`): pipeline parallelism, VLM, and LoRA fall back to full projection, and per-model support is decided by a bit-exact one-row probe rather than an allowlist — so head-side scaling quirks (Cohere `logit_scale`, Granite `logits_scaling`) degrade safely instead of silently changing logits.

The one place that still materializes full-width logits and then slices is the **draft-model ingest forward** (`draft_model_proposer.py`): each chunk projects every row and only `mx.take`s each plan's final row. That path already has chunked dispatch (`VLLM_METAL_SPEC_INGEST_CHUNK`) bounding the peak allocation, which is the same concern #482 direction 3 tracks — the symmetric selective projection there would be a further optimization, not part of this RFC's text-prefill scope.

Given the target-side behavior described in the RFC is in main with the safety gates above, this looks closable; the remaining draft-ingest work can continue under #482.
